### PR TITLE
Support geo.

### PR DESCRIPTION
**Pull request by CPWstatic**
Since we haven't implemented the index yet, I'll implement the **geo** through **multi-space** as we designed this on October 8th.
1. We will need a space that stores all the geographic location information.
2. We will find the locations through function **near**.

http://s2geometry.io/ 
We can refer to this website to understand s2geo.

### CODE REVIEW

**Comment by CPWstatic** [comment]
Jenkins go!

**Comment by CPWstatic** [comment]
Jenkins go!

**Comment by CPWstatic** [comment]
Jenkins go!

**Comment by laura-ding** [comment]
jenkins go

File: src/common/filter/geo/GeoParams.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+constexpr char kWktLinestringPrefix[] = "LINESTRING";
+constexpr char kWktPolygonPrefix[] = "POLYGON";
+
+constexpr double kEarthRadiusMeters = (6371.000 * 1000);
+
+struct RegionCoverParams {
+public:
+    S2RegionCoverer::Options regionCovererOpts() const {
+        S2RegionCoverer::Options opts;
+        opts.set_min_level(minCellLevel_);
```
Comment (laura-ding) [inline src/common/filter/geo/GeoParams.h]: If set it a negative number?

File: src/graph/test/GeoTest.cpp [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+    {
+        cpp2::ExecutionResponse resp;
+        std::string cmd = "CREATE EDGE locate(name string)";
+        auto code = client_->execute(cmd, resp);
+        if (cpp2::ErrorCode::SUCCEEDED != code) {
+            return TestError() << "Do cmd:" << cmd << " failed";
+        }
+    }
+
+    sleep(12);
```
Comment (laura-ding) [inline src/graph/test/GeoTest.cpp]: Why sleep so long time?

File: src/common/filter/geo/GeoFilter.cpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+#include "filter/geo/GeoIndex.h"
+#include <s2/s2cell_id.h>
+#include <s2/s2latlng.h>
+#include <s2/s2cap.h>
+#include <s2/s2region_coverer.h>
+
+namespace nebula {
+namespace geo {
+StatusOr<std::string> GeoFilter::near(const std::vector<VariantType> &args) {
+    auto predicate = args[0];
```
[the remark this answers is not in the document]
Comment (CPWstatic) [reply src/common/filter/geo/GeoFilter.cpp]: FuntionManager will guaranty its not empty.

File: src/common/filter/FunctionManager.cpp [hunk 4]
```diff
@@ -485,6 +486,47 @@ FunctionManager::FunctionManager() {
...
+    {
+        auto &attr = functions_["cos_similarity"];
+        attr.minArity_ = 2;
+        attr.maxArity_ = INT64_MAX;
+        attr.body_ = [] (const auto &args) {
+            if (args.size() % 2 != 0) {
+                LOG(ERROR) << "The number of arguments must be even.";
+                // value range of cos is [-1, 1]
+                // it means error when we return -2
+                return static_cast<double>(-2);
```
Comment (dutor) [inline src/common/filter/FunctionManager.cpp]: FYI. You could just `return -2lf`.

File: src/graph/GoExecutor.cpp [hunk 5]
```diff
@@ -179,6 +179,18 @@ Status GoExecutor::prepareFrom() {
...
                 break;
             }
+            if (expr->isFunCallExpression()) {
+                auto *funcExpr = static_cast<FunctionCallExpression*>(expr);
+                if (*(funcExpr->name()) == "near") {
+                    auto v = Expression::asString(value.value());
+                    std::vector<VertexID> result;
+                    folly::split(",", v, result, true);
+                    starts_.insert(starts_.end(),
+                                   std::make_move_iterator(result.begin()),
```
Comment (dutor) [inline src/graph/GoExecutor.cpp]: FYI, It won't have any benefit to move from trivial types.

**Pull request merged**